### PR TITLE
Fix isUnitLikeType to (again) handle tagged literal types

### LINKED ISSUE #51538: [4.9.2-rc regression] if/else narrowing with a branded string discriminant no longer eliminates matched constituents in later branches
# Bug Report

### 🔎 Search Terms

assert never branded string literal union type narrowing

### 🕗 Version & Regression Information

This changed between versions 4.8.4 and 4.9.2-rc

### ⏯ Playground Link

[Playground link](https://www.typescriptlang.org/play?ts=4.9.2-rc#code/C4TwDgpgBAMhBmwoF4oHIA2DhoNwChRIoAlASwHMALJVNAJ0prSgDIoBvKR64ALnQ9mUAL4Ei0AKJlgVCPRSxsUAD6kmwAvngBXAHYBjYGQD2eqAEMAzlfnAAchABu8gBROBe5-ICUn7woc+FBQsvQmAO5QXlGS9OH0rmheLvRoPgQi+PgAJhAGGBb00AZmVkhOFhg6EALSsvJaZPBQ7lU1KMh0WIjpnMFQpXrlUPAmJgKY2CyoldUQmVAQGLZQza1zHV10Qjg+-SFDIwBGRZO7M1CbC1AA9LeDJvH5wBgg0UXhERA5UHlWYBk0GO9AsehyZD0FAAdLD8CIlitoEEQtZbPQHAE2vMMncHrDoccdEg9CYkMsyABbSEWYA-SxWSyDMEQnK0iDwoA)

### 💻 Code

```ts
type Left = 'left';
type Right = 'right' & { right: 'right' };
type Either = Left | Right;

function assertNever(v: never): never {
  throw new Error('never');
}

declare const value: Either;

if (value === 'left') {
  const foo: 'left' = value;
} else if (value === 'right') {
  const bar: 'right' = value; // correctly narrowed despite branding...
} else {
  assertNever(value); // ...but not eliminated as a candidate
}
```

<details><summary><b>Output</b></summary>

```ts
"use strict";
function assertNever(v) {
    throw new Error('never');
}
if (value === 'left') {
    const foo = value;
}
else if (value === 'right') {
    const bar = value; // correctly narrowed despite branding...
}
else {
    assertNever(value); // ...but not eliminated as a candidate
}

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "strict": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "target": "ES2017",
    "jsx": "react",
    "module": "ESNext",
    "moduleResolution": "node"
  }
}
```

</details>
      
      

### 🙁 Actual behavior

The compiler claims the `assertNever` is reachable (by virtue of only narrowing `value` to `Right` in the last branch).

### 🙂 Expected behavior

The `assertNever` is considered unreachable (that is, `value` is narrowed to `never`).

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #51538.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test faster
@typescript-bot test top100

**Comment by ahejlsberg** [comment]
Tests and performance are unaffected. This should be good to merge.

**Review by RyanCavanaugh — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
I'll try, but this probably can't be automatically cherry-picked.

@typescript-bot cherry-pick this to release-4.9

It will probably fail - if it does, would you be able to send a PR with the test and change for `release-4.9`?

File: tests/cases/conformance/types/unknown/unknownControlFlow.ts [hunk 1]
```diff
@@ -429,3 +429,25 @@ type AB = "A" | "B";
...
+
+type Left = 'left';
+type Right = 'right' & { right: 'right' };
+type Either = Left | Right;
+
+function assertNever(v: never): never {
+    throw new Error('never');
+}
+
+function fx20(value: Either) {
```
Comment (DanielRosenwasser) [inline tests/cases/conformance/types/unknown/unknownControlFlow.ts:443]: I feel like I'm having deja vu and I've asked this question, but why is this test in a file called `unknownControlFlow.ts`?

Comment (ahejlsberg) [reply tests/cases/conformance/types/unknown/unknownControlFlow.ts:443]: You did indeed: https://github.com/microsoft/TypeScript/pull/51076#discussion_r998620628. The test is in that file for the same reason.

**Comment by DanielRosenwasser** [comment]
Pick is in 4.9 with b4d382b9b12460adf2da4cc0d1429cf19f8dc8be.

**Pull request merged**